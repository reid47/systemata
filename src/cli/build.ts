export function build(config: any) {
  console.log('building!');
  console.log(JSON.stringify(config));
}
