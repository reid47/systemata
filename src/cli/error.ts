export const error = (message: string) => {
  console.error(`${message}\n\nUse '-h' or '--help' to show usage information.`);
  process.exit(1);
};
