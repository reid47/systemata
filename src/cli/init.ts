import * as readline from 'readline';
import * as fs from 'fs';
import * as path from 'path';
import { Config } from '../types';
import { resolveConfig } from '../config';

const configFileTemplate = (config: Config) => `module.exports = {
  system: {
    name: ${JSON.stringify(config.system.name)},
    version: ${JSON.stringify(config.system.version)},
    description: ${JSON.stringify(config.system.description)}
  },
  colors: {},
  spacing: {}
};`;

const prompts = [{ text: 'System name', key: 'name' }, { text: 'Initial version', key: 'version' }];

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

function ask(config: Config, done: (config: Config) => void, promptIndex = 0) {
  const prompt = prompts[promptIndex];

  rl.question(`${prompt.text} (default: '${config.system[prompt.key]}'): `, answer => {
    if (answer) config.system[prompt.key] = answer;

    if (promptIndex === prompts.length - 1) {
      rl.close();
      done(config);
    } else {
      ask(config, done, promptIndex + 1);
    }
  });
}

function writeFile(configFile: string, contents: string) {
  const filePath = path.resolve(process.cwd(), configFile);

  try {
    fs.writeFileSync(filePath, contents);
    console.log(`Config file written: ${filePath}`);
  } catch (err) {
    console.error(`Error creating config file: ${err.message}`);
    process.exit(1);
  }
}

export function init(configFile: string, skipQuestions: boolean) {
  const defaultConfig = resolveConfig({});

  if (skipQuestions) {
    const output = configFileTemplate(defaultConfig);
    return writeFile(configFile, output);
  }

  ask(defaultConfig, userConfig => {
    const output = configFileTemplate(userConfig);
    writeFile(configFile, output);
  });
}
