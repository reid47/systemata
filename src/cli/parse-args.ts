import { CliInput } from '../types';

export const knownArgs: { [key: string]: string } = {
  '-h': 'help',
  '--help': 'help',
  '-v': 'version',
  '--version': 'version',
  '-c': 'configFile',
  '--config': 'configFile',
  '-y': 'skipQuestions',
  '--yes': 'skipQuestions'
};

const isFlag = (arg: string) => /^-[-]?[a-zA-Z]/.test(arg);

export function parseArgs(argv: string[]): CliInput {
  const args: { [key: string]: string | boolean } = {};

  let command: string | undefined;
  let lastArg: string = '';

  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i];

    if (i === 0 && !isFlag(arg)) {
      command = arg;
      continue;
    }

    if (knownArgs[arg]) {
      args[knownArgs[arg]] = true;
    } else if (lastArg && knownArgs[lastArg]) {
      args[knownArgs[lastArg]] = arg;
    }

    lastArg = arg;
  }

  return { args, command } as CliInput;
}
