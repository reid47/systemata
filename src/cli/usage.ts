import { name, version, description } from '../../package.json';

export const usage = `${name} (version ${version})
  ${description}

Usage:
  ${name} [options]
  ${name} init [init options]
  ${name} develop [develop options]
  ${name} build [build options]

Options:
  -h, --help              print this usage info
  -v, --version           print ${name} version

'init' options:
  -c, --config            path where config file will be created
  -y, --yes               skip questions, use default initial settings

'develop' options:
  -c, --config            path to config file

'build' options:
  -c, --config            path to config file
`.trim();
