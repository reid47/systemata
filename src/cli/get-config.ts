import * as path from 'path';
import { CliInput, Config } from '../types';
import { error } from './error';

export function getConfig(input: CliInput): Config | undefined {
  const { configFile } = input.args;
  const cwd = process.cwd();

  const possibleConfigFiles: string[] = [
    path.resolve(cwd, 'systemata.config.js'),
    path.resolve(cwd, 'systemata.config.json')
  ];

  if (typeof configFile === 'string') {
    possibleConfigFiles.unshift(path.resolve(cwd, configFile));
  }

  let config: Config | undefined;
  const errorMessages = [];

  for (let i = 0; i < possibleConfigFiles.length; i++) {
    try {
      config = require(possibleConfigFiles[i]);
      break;
    } catch (err) {
      errorMessages.push(err.message);
    }
  }

  if (!config) {
    error(`Could not load config file: ${errorMessages.join(' ')}`);
    return;
  }

  return config;
}
