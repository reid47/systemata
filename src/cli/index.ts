#!/usr/bin/env node
import * as path from 'path';
import * as fs from 'fs';
import { build } from './build';
import { version } from '../../package.json';
import { parseArgs, knownCommands } from './parse-args';
import { usage } from './usage';

const error = (message: string) => {
  console.error(`${message}\n\nUse '-h' or '--help' to show usage information.`);
  process.exit(1);
};

const { args, command } = parseArgs(process.argv.slice(2));

if (args.help) {
  console.log(usage);
  process.exit(0);
}

if (args.version) {
  console.log(version);
  process.exit(0);
}

if (command === 'init') {
  const configFile = args['-c'] || args['--config'];

  if (!configFile || typeof configFile !== 'string') {
    error("No config file specified. Use '-c' or '--config' to pass a path to a system config file.");
  }

  if (fs.existsSync(path.resolve(process.cwd(), configFile as string))) {
    error(`File '${configFile}' already exists. Please specify a file name that doesn't exist.`);
  }
}

if (command === 'build') {
  const configFile = args['-c'] || args['--config'];

  if (!configFile || typeof configFile !== 'string') {
    error("No config file specified. Use '-c' or '--config' to pass a path to a system config file.");
  }

  let config;
  try {
    config = require(path.resolve(process.cwd(), configFile as string));
  } catch (err) {
    error(`Error loading config file: ${err.message}`);
  }

  try {
    build(config);
    process.exit(0);
  } catch (err) {
    error(err);
  }
}

const validCommands = Object.keys(knownCommands).join(', ');

if (!command) {
  error(`No command specified. Command should be one of: ${validCommands}`);
}

error(`Invalid command specified: ${command}\nCommand should be one of: ${validCommands}`);
