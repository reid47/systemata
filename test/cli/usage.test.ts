import { usage } from '../../src/cli/usage';

test('main', () => {
  expect(usage.main).toMatchSnapshot();
});

test('archive', () => {
  expect(usage.archive).toMatchSnapshot();
});

test('build', () => {
  expect(usage.build).toMatchSnapshot();
});

test('develop', () => {
  expect(usage.develop).toMatchSnapshot();
});

test('init', () => {
  expect(usage.init).toMatchSnapshot();
});

test('show-config', () => {
  expect(usage['show-config']).toMatchSnapshot();
});
