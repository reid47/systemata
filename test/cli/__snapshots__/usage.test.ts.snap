// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`archive 1`] = `
"systemata archive, version 0.0.1
  Create a static, versioned copy of the current design system.

Usage:
  systemata archive [options]

Global options:
  -h, --help       Print usage information
  -v, --version    Print systemata version

'archive' options:
  -c, --config <configFile>   Path to config file"
`;

exports[`build 1`] = `
"systemata build, version 0.0.1
  Build all design system files

Usage:
  systemata build [options]

Global options:
  -h, --help       Print usage information
  -v, --version    Print systemata version

'build' options:
  -c, --config <configFile>   Path to config file"
`;

exports[`develop 1`] = `
"systemata develop, version 0.0.1
  Start a server that watches files for changes & rebuilds on change.

Usage:
  systemata develop [options]

Global options:
  -h, --help       Print usage information
  -v, --version    Print systemata version

'develop' options:
  -c, --config <configFile>   Path to config file"
`;

exports[`init 1`] = `
"systemata init, version 0.0.1
  Creates a new system config file.

Usage:
  systemata init [options]

Global options:
  -h, --help       Print usage information
  -v, --version    Print systemata version

'init' options:
  -c, --config <configFile>   Path to config file
  -y, --yes                   Skip questions, use defaults"
`;

exports[`main 1`] = `
"systemata, version 0.0.1
  Build a design system with one file

Usage:
  systemata <command> [options]

Global options:
  -h, --help       Print usage information
  -v, --version    Print systemata version

Commands:
  archive          Archive a static copy of current design system
  build            Build all design system files
  develop          Start a build server & re-build on change
  init             Create a new design system

For help with a command, use 'systemata <command> -h'."
`;

exports[`show-config 1`] = `
"systemata show-config, version 0.0.1
  Prints resolved system config as systemata sees it. Useful for making sure configuration looks as you expect.

Usage:
  systemata show-config [options]

Global options:
  -h, --help       Print usage information
  -v, --version    Print systemata version

'show-config' options:
  -c, --config <configFile>   Path to config file"
`;
